proposals/interface-types/working-notes/record.md

```diff
@@ -0,0 +1,73 @@
+
+
+### Records
+
+Records are collections of named and typed fields. Records can be defined in as
+Interface Types; for example this definition encodes a familiar notion of a
+credit card:
+
+```wat
+(@interface datatype @cc 
+  (record "cc"
```

Comment:
We should discuss whether it's useful to add string names to type definitions.  I think it's necessary to name fields (since it allows specifying field type compatibility in an order-independent manner, which is useful), but I don't think it buys us anything to name structs, given that we'll already need to check their contents for type-compatibility anyway.

Reply:
If you have alternate cases in an algebraic type then you need to distinguish them. This is going to be an issue especially for C ... to be handled later.
It boils down to a couple of issues: there are certain specific cases which are very common and which are technically instances of algebraic type definitions - do we wish to honor that heritage; and there are also legitimate API scenarios where there is variation in the data being passed.
The first is essentially the option type and also the either type (used to distinguish normal from error cases).
The second includes supporting the kind of polymorphism that is quite common in the wild, and in JS APIs. 
Having different cases allows us to capture the variants that show up in APIs without inventing new names where there weren't before: e.g., 

addEventListener('type', function, <Boolean> capture?);
addEventListener('type', function, <Object> options?);
addEventListener('type', Object handler, <Boolean> capture?);
addEventListener('type', Object handler, <Object> options?);

Reply:
Interesting point.  With the current interface types proposal, one never does an "overload resolution" wherein a (name, signature) pair is looked up in some scope for a function.  Rather, the order is inverted: a function is supplied to wasm instantiation and then we ask if that function's type is compatible with the import's declared type.  For Web IDL, in fact all declared overloads end up producing a single function that wasm can import and thus the type-compatibility check is against this list of valid signatures.  For future APIs (like WASI), I think we'd want to instead not allow overloading and instead use variant types in the signature (which, of course, have string field names).